src/keys/mod.rs

```diff
@@ -0,0 +1,120 @@
+pub mod byte_key;
+pub mod nibble_key;
+
+pub use byte_key::*;
+pub use nibble_key::*;
+
+/// An iterator that will walk the elements of the key. `U` is the unit
+/// type of a key element (e.g. byte, bit, nibble) and `K` is the key type.
+pub struct KeyIterator<'a, U, K>
+where
+    U: Copy + PartialEq + std::fmt::Debug + Default,
+    K: Key<U> + std::ops::Index<usize, Output = U>,
+{
+    /// Index of the element that the iterator is currently pointing at.
+    item_num: usize,
+
+    /// A reference to the object being iterated.
+    container: &'a K,
+
+    element: U,
+}
+
+impl<'a, U, K> std::fmt::Debug for KeyIterator<'a, U, K>
+where
+    U: std::marker::Sized + Copy + PartialEq + std::fmt::Debug + Default,
+    K: Key<U> + std::ops::Index<usize, Output = U>,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "iterator at element {}/{}",
+            self.item_num,
+            self.container.len()
+        )
+    }
+}
+
+impl<'a, U, K> Iterator for KeyIterator<'a, U, K>
+where
+    U: std::marker::Sized + Copy + PartialEq + std::fmt::Debug + Default,
+    K: Key<U> + std::ops::Index<usize, Output = U>,
+{
+    type Item = U;
+
+    fn next(&mut self) -> Option<U> {
+        if self.item_num < self.container.len() {
+            let element: U = self.container[self.item_num];
+            self.item_num += 1;
+            self.element = element;
+            Some(element)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a, U, K> KeyIterator<'a, U, K>
+where
+    U: std::marker::Sized + Copy + PartialEq + std::fmt::Debug + Default,
+    K: Key<U> + std::ops::Index<usize, Output = U>,
+{
+    fn rewind(&mut self) {
+        assert!(self.item_num > 0);
+        self.item_num -= 1;
+    }
+
+    /// Compares the two iterators and leave them at their first differing
+    /// element.
+    pub fn chop_common(&mut self, other: &mut Self) -> usize {
+        loop {
+            let (rewind, brk) = match (self.next(), other.next()) {
+                // Both iterators reached the end, keys are
+                // identical.
+                (None, None) => (false, true),
+                // One of the iterators has reached the end,
+                // advance both and return.
+                (_, None) | (None, _) => (true, true),
+                // Both iterators are still pointing at a
+                // value, check if these values are the same.
+                // If they are, loop, otherwise quit.
+                (Some(x), Some(y)) => (x != y, x != y),
+            };
+
+            if rewind {
+                self.rewind();
+                other.rewind();
```

Comment:
@Schaeff because the calling code expects the next call to one iterator to return the first differing byte in the key, I find myself having to rewind the iterator by one byte. I have tried to use `Peekable` but it will also advance the iterator underneath. Any suggestion here?

Reply:
Then the comment at line 67 is incorrect as you're leaving them at their last common element?

From [Peekable](https://doc.rust-lang.org/std/iter/struct.Peekable.html) it seems that your implementation of the trait was incorrect as it's not supposed to advance the iterator, maybe an issue there?

Reply:
The comment is indeed too vague. The thing is: `Peekable` is an iterator, that doesn't advance on `peek`. But the iterator that it consumes *does* advance. It's not a problem with just this iterator, it's the problem for all of them:

```rust
fn main() {
    let v = vec![1, 2, 3];
    let mut pit = v.iter().peekable();
    println!("before peek {:?}", pit);
    pit.peek();
    println!("after peek {:?}", pit);
}
```

output:

```
before peek Peekable { iter: Iter([1, 2, 3]), peeked: None }
after peek Peekable { iter: Iter([2, 3]), peeked: Some(Some(1)) }
```

Reply:
Also, using Peekable isn't great for another reason: since it's not defined in the same package as `KeiIterator`, I wouldn't be able to perform the trick that saves me a copy :disappointed:

Reply:
Okay my bad it looks more complicated than that.